### ISSUE TITLE
Simple mis-spelled terms do not provide the correct results.

### ISSUE THREAD

**Issue by muffinresearch** [mozilla/addons#798]
Filed as a good example from the PRD:

If you want to find “Tabby Cat,” and you search “Tabbycat,”  “Tabbbycat,” or “Tabbicat” you will not get “Tabby Cat” in the first page of results.

- [Tabbycat](https://addons.mozilla.org/en-US/firefox/search/?platform=mac&q=tabbycat)
- [Tabbbycat](https://addons.mozilla.org/en-US/firefox/search/?platform=mac&q=tabbbycat)
- [Tabbicat](https://addons.mozilla.org/en-US/firefox/search/?platform=mac&q=tabbicat)

Another similar example would be:

If you want to find the extension “Tree Style Tab” (three words) and erroneously search “Treestyle Tab” (two words), the extension does not appear until the 10th result: this could be a relevance issue? I tested “Threestyle Tab” and got the right result.
“[Treestyle Tab](https://addons.mozilla.org/en-US/firefox/search/?platform=mac&q=Treestyle+Tab)”

**Comment by diox** [comment]
With our current configuration, the following things have an impact depending on what you're searching for:
- The [dictionary decompounder](https://www.elastic.co/guide/en/elasticsearch/reference/5.6/analysis-compound-word-tokenfilter.html#_dictionary_decompounder). It decompounds words into multiple tokens according to a dictionary. That dictionary needs to stay limited for performance reasons.
- The [fuzzy query](https://www.elastic.co/guide/en/elasticsearch/reference/5.6/query-dsl-fuzzy-query.html) on the name field. It's configured with a prefix of 4, and a `fuzzyness` of `AUTO` - the actual value depends on the length of the query. 
- The [prefix query](https://www.elastic.co/guide/en/elasticsearch/reference/5.6/query-dsl-prefix-query.html) on the name (when you search for "something" and the add-on is named "somethingawesome")
- The regular match query on the description (sometimes the words we're looking for appear separated by spaces in there, which helps)

For `tree style tab` I already have a fix, in https://github.com/mozilla/addons-server/issues/9791 while documenting the `dictionary_decompounder` I'll add "style" and "tree" to the dictionary, it should improve things a ton.

For the more general case, I think we'll have to play with other analyzers or fuzzy searching.

**Comment by diox** [comment]
https://github.com/mozilla/addons/issues/567 decreases the prefix length for fuzzy matching from 4 to 2, so that should help typos that occur in the beginning of each term (except on the first 2 chars).

Here is what we can play with after that:
- Improve partial matching with [n-grams](https://www.elastic.co/guide/en/elasticsearch/reference/5.6/analysis-ngram-tokenizer.html) and [shingles](https://www.elastic.co/guide/en/elasticsearch/reference/5.6/analysis-shingle-tokenfilter.html)
- Do [spelling corrections](https://www.elastic.co/guide/en/elasticsearch/reference/5.6/analysis-hunspell-tokenfilter.html) when analyzing text
- Relax [fuzzy matching](https://www.elastic.co/guide/en/elasticsearch/reference/5.6/common-options.html#fuzziness) even more, but that have too big of a performance penalty
- Add [synonyms](https://www.elastic.co/guide/en/elasticsearch/reference/5.6/analysis-synonym-tokenfilter.html) from a dictionary (doesn't help that much with typos/misspells though)
- Implement [suggesters](https://www.elastic.co/guide/en/elasticsearch/reference/5.6/search-suggesters.html) for a "Did you mean?" kind of response. This would require frontend and UX work

**Comment by ioanarusiczki** [comment]
@diox 

I tested with FF63 (Win10) and here are my test results:
- no problems with Tree Style Tab, I tried all kind of variations and got it suggested and displayed as the first result. 

![tree style](https://user-images.githubusercontent.com/33448286/49451486-f2559080-f7e7-11e8-8a8f-c63ce286a203.png)

- the "Tabby cat" theme on AMO -dev tried with "tabbycat" or "tabbbycat" is placed on the third page of results. "Tabbicat" is on page 6. If I try searching as "tabbycat", "tabbbycat" and "tabicat" on AMO prod they're all on page 33/34 so I think there has been an improvement made. 
Other names do not seem to have this problem when they're searched so this should be an isolated case.

- what I've also noticed is that there's a "tabby cat" extension on AMO prod which searched with "tabbycat" and the other variations it is still suggested on the first page among the results (the default sort by filter is "relevance" and the extension has 4,131 users). So, I submitted a "tabby cat" extension on AMO dev and tried to search it as "tabbycat" or tabbbycat" and I found it on page 16 (meanwhile the "tabby cat" theme was still on page 3).  Again,  perhaps the "Relevance" sort by filter (which is the default) will prioritize the search results after the number of users too.  My "tabby cat" extension on -dev has 0 users and the "tabby cat" theme on dev has 2 users.  

"tabby cat" searched as "tabbycat" the extension and theme on AMO dev

![tabby cat dev](https://user-images.githubusercontent.com/33448286/49453772-73af2200-f7ec-11e8-8eaa-4078b5deca9a.gif)


Let me know if I can mark the issue verified.
Thanks!

**Comment by diox** [comment]
Number of users does matter a little bit, yes. I've lowered the impact, but it still matters. We'll re-evaluate once it lands on prod, at least this hasn't made the situation worse :)

**Comment by ioanarusiczki** [comment]
@diox Yes, good idea. This must be verified again in production and see the impact there.

**Comment by ioanarusiczki** [comment]
@diox I verified the Tabby cat again on AMO prod and the results are:
- "Tabby Cat" the extension is on the first page on the list of suggestions no matter how I searched for it "tabbbycat", "tabbycat" or "tabicat"
- "tabby cat" the theme is on page 25 when searched as "tabbycat", page 28 when searched as "tabbycat" and I did not find it in the first 40 pages when I tried with "tabicat".  Also before it there is "Beautiful tabby cat" (10 users) or other themes with similar names. 
On AMO prod "tabby cat" theme has 3 users , the extension has 4315 users.

- I think "Tree Style Tab" looks good
![onetree](https://user-images.githubusercontent.com/33448286/50011143-f8a3f380-ffc3-11e8-910c-6e5ddfcb8cf4.png)

![two trees](https://user-images.githubusercontent.com/33448286/50011159-06597900-ffc4-11e8-9142-3bc2dd77b275.png)

Let me know what you think.
